### PR TITLE
[rllib] General RNN support

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

This adds support for RNNs. It can be enabled on PG / A3C by setting `{"model": {"use_lstm": True, "max_seq_len": N}}`, which will wrap the model in a LSTM cell that is trained with max seq len N.

The main trick is that we add the time dimension at the last moment. The non-LSTM layers of the model see their inputs as one flat batch. Before the LSTM cell, we reshape the input to add the expected time dimension. During postprocessing, we dynamically pad the experience batches so that this reshaping is possible. See lstm.py:`add_time_dimension` and `chop_into_sequences` for more information.

A3C's RNN support now goes through this path.

- [x] Wait for https://github.com/ray-project/ray/pull/2294 to merge
- [x] Benchmarks
- [x] Let's do PPO support in another PR (I think the main thing is to make sure sequences are divided up among the multiple gpus correctly)

Benchmark results:

Tested A3C and it doesn't seem to have regressed from master. However, perf overall seems ~2-3x slower than expected. I checked and this regression wasn't due to any of the recent refactoring though, so we can probably go ahead and merge this.

test

## Related issue number

https://github.com/ray-project/ray/issues/2240

### CODE REVIEW

File: python/ray/rllib/models/lstm.py [hunk 1]
```diff
@@ -2,56 +2,173 @@
...
+from ray.rllib.models.misc import linear, normc_initializer
 from ray.rllib.models.model import Model
 
 
+def add_time_dimension(padded_inputs, seq_lens):
+    """Adds a time dimension to padded inputs.
+
+    Arguments:
+        padded_inputs (Tensor): a padded batch of sequences. That is,
+            for seq_lens=[1, 2, 2], then inputs=[A, *, B, B, C, C], where
```
Comment (richardliaw) [inline python/ray/rllib/models/lstm.py]: When will you have situations where you pad in the middle of the given sequence? ie, only during batch SGD where your batch has N trajectories?

File: python/ray/rllib/utils/tf_policy_graph.py [hunk 2]
```diff
@@ -99,15 +111,30 @@ def compute_actions(
...
+
+        # Simple case
+        if not self._state_inputs:
+            for k, ph in self._loss_inputs:
+                feed_dict[ph] = batch[k]
+            return feed_dict
+
+        # RNN case
+        feature_keys = [
+            k for k, v in self._loss_inputs if not k.startswith("state_in_")]
```
Comment (richardliaw) [inline python/ray/rllib/utils/tf_policy_graph.py]: should add note that if someone bakes their own RNN model and wants to use this functionality, they need to prefix like this

File: python/ray/rllib/utils/tf_policy_graph.py [hunk 3]
```diff
@@ -34,25 +35,31 @@ class TFPolicyGraph(PolicyGraph):
...
             loss_inputs (list): a (name, placeholder) tuple for each loss
                 input argument. Each placeholder name must correspond to a
-                SampleBatch column key returned by postprocess_trajectory().
+                SampleBatch column key returned by postprocess_trajectory(),
+                and has shape [BATCH_SIZE, data...].
             is_training (Tensor): input placeholder for whether we are
                 currently training the policy.
             state_inputs (list): list of RNN state output Tensors.
             state_outputs (list): list of initial state values.
+            seq_lens (Tensor): placeholder for RNN sequence lengths.
```
Comment (richardliaw) [inline python/ray/rllib/utils/tf_policy_graph.py]: can you add shape for this tensor? they are very helpful

**Review by richardliaw — COMMENTED** [review]
questions and documentation

File: python/ray/rllib/models/lstm.py (hunk 1 above)
Comment (ericl) [reply python/ray/rllib/models/lstm.py]: Yes, in general you will have to pad when-ever you have variable length trajectories and they aren't all divisible by max_seq_len

File: python/ray/rllib/models/lstm.py [hunk 4]
```diff
@@ -2,56 +2,173 @@
...
+            i += l
+        initial_states.append(np.array(s_init))
+
+    return feature_sequences, initial_states, np.array(seq_lens)
+
+
 class LSTM(Model):
-    """Vision LSTM network based here:
-    https://github.com/openai/universe-starter-agent"""
+    """Adds a LSTM cell on top of some other model output.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/models/lstm.py]: Done.

File: python/ray/rllib/test/test_lstm.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/test/test_lstm.py]: Fixed

File: python/ray/rllib/utils/tf_policy_graph.py (hunk 3 above)
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: Done

File: python/ray/rllib/utils/tf_policy_graph.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: These are automatically added by sampler, so it's not needed.

File: python/ray/rllib/utils/tf_policy_graph.py [hunk 6]
```diff
@@ -99,15 +111,30 @@ def compute_actions(
...
+                feed_dict[ph] = batch[k]
+            return feed_dict
+
+        # RNN case
+        feature_keys = [
+            k for k, v in self._loss_inputs if not k.startswith("state_in_")]
+        state_keys = [
+            k for k, v in self._loss_inputs if k.startswith("state_in_")]
+        feature_sequences, initial_states, seq_lens = chop_into_sequences(
+            batch["t"],
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/utils/tf_policy_graph.py]: It's automatically added by sampler.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**